### PR TITLE
Divide isSub into Attributes

### PR DESCRIPTION
**Pull request by ron-murhammer**
## Overview
- Second part of https://forums.triplea-game.org/topic/328/unit-option-can-submerge-hide-for-land-units-partisan-guerrilla-spy-diplomat-munition
- Breaking isSub and isDestroyer into smaller attributes so more fine grained control can be achieved (canHide, firstStrike, canTarget, canBeTargetedBy, etc)
- This should have no impact on existing maps and sea subs. Most functionality will now work for land/air subs but there are still some further changes need especially for cantTarget and cantBeTargetedBy as these are hardcoded to just check if empty or not.

## Functional Changes
**Separate isSub Functions**
1. **canEvade** - submerge
2. **isFirstStrike** - roll and take casualties before regular units
3. **canNotTarget** - can't target specific units like air units
4. **canNotBeTargetedBy** - can't be targeted by specific units like air units
5. **canMoveThroughEnemies** - unblockable, Treat Hostile Territories as Friendly
6. **canBeMovedThroughByEnemies** - enemies can move through territories if only these types of units are present (would also need blitz for enemy land territories)

## Manual Testing Performed
- Launched a few maps including revised, ww2v3, and global to ensure games still work properly

### CODE REVIEW

File: game-core/src/main/java/games/strategy/triplea/delegate/Matches.java [hunk 1]
```diff
@@ -106,12 +106,36 @@ private Matches() {}
     return unit -> UnitAttachment.get(unit.getType()).getIsSea();
   }
 
-  public static Predicate<Unit> unitIsSub() {
-    return unit -> UnitAttachment.get(unit.getType()).getIsSub();
+  // public static Predicate<Unit> unitIsSub() {
+  // return unit -> UnitAttachment.get(unit.getType()).getIsSub();
+  // }
```
Comment (RoiEXLab) [inline game-core/src/main/java/games/strategy/triplea/delegate/Matches.java]: This commented out code should be removed

File: game-core/src/main/java/games/strategy/triplea/delegate/Matches.java [hunk 2]
```diff
@@ -106,12 +106,36 @@ private Matches() {}
...
+
+  public static Predicate<Unit> unitCanMoveThroughEnemies() {
+    return unit -> UnitAttachment.get(unit.getType()).getCanMoveThroughEnemies();
+  }
+
+  public static Predicate<Unit> unitCanBeMovedThroughByEnemies() {
+    return unit -> UnitAttachment.get(unit.getType()).getCanBeMovedThroughByEnemies();
+  }
+
+  public static Predicate<Unit> unitCantTargetAll() {
```
Comment (MelvinWM) [inline game-core/src/main/java/games/strategy/triplea/delegate/Matches.java]: Is this a duplicate of `unitCantBeTargetedByAll`?

File: game-core/src/main/java/games/strategy/triplea/delegate/Matches.java [hunk 3]
```diff
@@ -106,12 +106,36 @@ private Matches() {}
...
+    return unit -> !UnitAttachment.get(unit.getType()).getCantBeTargetedBy().isEmpty();
```
Comment (MelvinWM) [inline game-core/src/main/java/games/strategy/triplea/delegate/Matches.java]: Should this be?:

    return unit -> !UnitAttachment.get(unit.getType()).getCantTarget().isEmpty();

**Comment by MelvinWM** [comment]
I have only looked at the changes somewhat superficially, but they also look (very) good to me.

File: game-core/src/main/java/games/strategy/triplea/delegate/Matches.java (hunk 3 above)
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/Matches.java]: Nice catch. You are correct. Its fixed now.

File: game-core/src/main/java/games/strategy/triplea/delegate/Matches.java (hunk 1 above)
Comment (ron-murhammer) [reply game-core/src/main/java/games/strategy/triplea/delegate/Matches.java]: Yep. Removed.

**Comment by ron-murhammer** [comment]
Thanks for the quick reviews. Updated to fix the 2 issues that were pointed out.

**Pull request merged**